SYCL/ESIMD/api/functional/ctors/common.hpp

```diff
@@ -0,0 +1,165 @@
+//===-- common.hpp - This file provides common functions for simd constructors
+//      tests -------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// Common file for test on simd constructors.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/experimental/esimd.hpp>
+
+#include "../../../esimd_test_utils.hpp"
+#include "../logger.hpp"
+#include "../type_coverage.hpp"
+
+#include <type_traits>
+
+namespace esimd_test {
+namespace api {
+namespace functional {
+
+// Dummy kernel for submitting some code into device side.
+template <typename DataT, int NumElems, typename T> struct Kernel;
+
+template <typename DataT>
+using allocator_type = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+
+template <typename DataT>
+using vector_with_allocator = std::vector<DataT, allocator_type<DataT>>;
+
+// Struct that provise methods for constructing inf, lowest, nan and max values
+// for reference data.
+template <typename DataT> struct value {
+  static DataT inf() {
+    DataT inf_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      inf_val = 0;
+    } else {
+      inf_val = std::numeric_limits<DataT>::infinity();
+    }
+    return inf_val;
+  }
+
+  static DataT lowest() {
+    DataT min_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      min_val = 0;
+    } else {
+      min_val = std::numeric_limits<DataT>::lowest();
+    }
+    return min_val;
+  }
+
+  static DataT nan() {
+    DataT nan_val{};
+    if constexpr (std::is_same_v<DataT, double>) {
+      nan_val = sycl::nan(42UL);
+    } else if constexpr (std::is_same_v<DataT, float>) {
+      nan_val = sycl::nan(42U);
+    } else if constexpr (std::is_same_v<DataT, sycl::half>) {
+      nan_val = 0;
+    }
+    return nan_val;
+  }
+
+  static DataT max() {
+    DataT max_val{};
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      max_val = 0;
+    } else {
+      max_val = std::numeric_limits<DataT>::max();
+    }
+    return max_val;
+  }
+};
+
+// Generate std::vector with reference according to current tested data type and
+// number of elements.
+template <typename DataT, int NumElems>
+void generate_ref_data(vector_with_allocator<DataT> &ref_data) {
```

Comment:
```suggestion
vector_with_allocator<DataT> generate_ref_data() {
```
So it would be possible to optimize allocation using `std::vector::reserve()` or appropriate constructor call.

Reply:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L128-R121)